File: android/src/main/java/expo/modules/brownfield/Messaging.kt
package expo.modules.brownfield

import java.util.UUID

typealias BrownfieldMessage = Map<String, Any?>
typealias BrownfieldCallback = (BrownfieldMessage) -> Unit
typealias BrownfieldFilter = ((BrownfieldMessage) -> Boolean)?

object BrownfieldMessaging {
  data class BrownfieldListener(
    val id: String,
    val filter: BrownfieldFilter = null,
    val callback: BrownfieldCallback
  )

  private val listeners = mutableSetOf<BrownfieldListener>()
  private var expoModule: ExpoBrownfieldModule? = null

  fun addListener(
    filter: BrownfieldFilter = null,
    callback: BrownfieldCallback
  ): String {
    val id = java.util.UUID.randomUUID().toString()
    listeners.add(BrownfieldListener(
      id,
      filter,
      callback
    ))

    return id
  }

  fun removeListener(id: String) {
    listeners.removeAll { it.id == id }
  }

  fun sendMessage(message: BrownfieldMessage) {
    expoModule?.let { module ->
      print(message)
      module.sendMessage(message)
    }
  }

  internal fun emit(message: BrownfieldMessage) {
    listeners.forEach { listener ->
      if (listener.filter?.invoke(message) != false) {
        listener.callback(message)
      }
    }
  }

  internal fun setExpoModule(expoModule: ExpoBrownfieldModule?) {
    this.expoModule = expoModule
  }
}

File: ios/Messaging.swift
import Foundation

public typealias BrownfieldMessage = [String: Any?]
public typealias BrownfieldCallback = (BrownfieldMessage) -> Void

public class BrownfieldMessagingInternal {
  public static let shared = BrownfieldMessagingInternal()

  private var listeners: [String: BrownfieldCallback] = [:]
  private var expoModule: ExpoBrownfieldModule? = nil

  private init() {}

  @discardableResult
  public func addListener(
    _ callback: @escaping BrownfieldCallback
  ) -> String  {
    let id = UUID().uuidString
    listeners[id] = callback
    return id
  }

  public func removeListener(id: String) {
    listeners.removeValue(forKey: id)
  }

  public func sendMessage(_ message: BrownfieldMessage) {
    expoModule?.sendMessage(message)
  }

  internal func emit(_ message: BrownfieldMessage) {
    listeners.values.forEach { listener in
      listener(message)
    }
  }

  internal func setExpoModule(_ expoModule: ExpoBrownfieldModule?) {
    self.expoModule = expoModule
  }
}
